pkg/logging/http.go

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package logging
+
+import (
+	"fmt"
+
+	"github.com/go-kit/kit/log"
+
+	"net/http"
+	"time"
+	"github.com/thanos-io/thanos/pkg/server/http/middleware"
+)
+
+type LogServerMiddleware struct {
+	opts   *options
+	logger log.Logger
+}
+
+// Things for logger
+// * request-id
+// * err code mapping to level
+// * duration of request
+
+func (c *LogServerMiddleware) HTTPMiddleware(name string, next http.Handler) http.HandlerFunc {
+
+	return func(w http.ResponseWriter, r *http.Request) {
+
+		// If request logging is disabled
+		if !c.opts.shouldLog(name) {
+			next.ServeHTTP(w, r)
+			return
+		}
+
+		logger := c.logger
+		logger = log.With(logger, "HTTP[Server]", name)
+		start := time.Now()
+
+		if id, ok := middleware.RequestIDFromContext(r.Context()); ok {
+			logger = log.With(logger, "request-id", id)
+		} else {
+			logger = log.With(logger, "request-id", "NONE")
+		}
+
+		wrapped := wrapResponseWriter(w)
+		next.ServeHTTP(wrapped, r)
+
+		status := wrapped.Status()
+		// server error, log them
+		if status >= 500 && status < 600 {
```

Comment:
What about using **exactly the same decider logic** as for gRPC?

Reply:
This is to simplify! :hugs:

Reply:
We can take care of this in another PR, too big of a change to be covered in this PR